### PR TITLE
Winston 3 two

### PR DESCRIPTION
**Pull request by mattcollier**

### CODE REVIEW

File: lib/loggers/index.js
Comment (mattcollier) [inline lib/loggers/index.js:74]: I think it is safe to just to this on L74 even if `meta.length === 0`, and thus avoid the special casing.  See:

https://repl.it/@mattcollier/OfficialFlippantTechnologies#index.js

File: lib/loggers/index.js
Comment (mattcollier) [inline lib/loggers/index.js:85]: @dlongley createChild does this which I think is just weird.  I don't think we need it.  I think making a copy and setting `defaultData` is adequate.  What do you think?

https://github.com/winstonjs/winston/blob/master/lib/winston/logger.js#L45-L71

File: lib/loggers/index.js
Comment (mattcollier) [inline lib/loggers/index.js:69]: @dlongley `defaultMeta` is added in winston here: https://github.com/winstonjs/winston/blob/master/lib/winston/create-logger.js#L80

You comfortable with simplifying this in the single argument case?

File: lib/loggers/index.js
Comment (dlongley) [reply lib/loggers/index.js:74]: Agreed.

File: lib/loggers/index.js
Comment (dlongley) [reply lib/loggers/index.js:85]: Yeah, I saw that it was doing special things with some `write` function, and I didn't know where/how it was used. That's why I kept it in there, to minimize changes. Presumably, `write` is used somewhere ... and it may break if we don't just keep the default behavior. I don't think it really hurts us to keep it.

File: lib/loggers/index.js
Comment (dlongley) [reply lib/loggers/index.js:69]: Yeah, I'm fine if it works without it. I'd rather reduce changes to the lib as much as possible.

File: lib/config.js
Comment (dlongley) [inline lib/config.js:73]: Are these going to be breaking changes? I was hoping to avoid this being a breaking release.

File: lib/loggers/formatters.js
Comment (dlongley) [inline lib/loggers/formatters.js:25]:
```suggestion
  const stringifiedRest = jsonStringify({
    ...info,
    level: undefined,
    message: undefined,
    module: undefined,
    splat: undefined,
    timestamp: undefined,
    workerId: undefined,
    workerPid: undefined,
  }, null, 2);
```

File: lib/loggers/formatters.js
Comment (dlongley) [inline lib/loggers/formatters.js:46]: Just flagging this for removal.

File: lib/loggers/index.js
Comment (dlongley) [inline lib/loggers/index.js:38]: I guess these all got reversed in later versions of winston.

File: lib/loggers/index.js
Comment (dlongley) [inline lib/loggers/index.js:55]:
```suggestion
```

File: lib/loggers/index.js
Comment (dlongley) [inline lib/loggers/index.js:165]: I'm just going to presume the above changes are winston 3.x equivalents.

File: lib/loggers/index.js
Comment (dlongley) [inline lib/loggers/index.js:314]: I am a little concerned we're losing some kind of functionality or corner case handling here. Maybe it's fine. Your thoughts, @mattcollier?

**Review by dlongley — APPROVED** [review]
Seems like this is well on its way. Just a few questions/responses to your questions. Approving ahead of time if you feel like my questions are/will be adequately addressed.

File: lib/loggers/index.js
Comment (dlongley) [reply lib/loggers/index.js:314]: Maybe this is all handled by the formatters/fast JSON stringify stuff I saw elsewhere.

File: lib/loggers/index.js
Comment (mattcollier) [reply lib/loggers/index.js:85]: OK, the `.child` implementation (referenced above) accepts only one parameter.  I'll be removing the `...args` bit.

File: lib/config.js
Comment (mattcollier) [reply lib/config.js:73]: @dlongley I was building this out with the intention of it being a breaking change.  The most important reason why this is the case is the fact that the Bedrock API is designed such that configuration options are being sent directly into Winston.  For instance `config.loggers.console` is passed directly into the `console` logger.  There have been many breaking changes in the Winston configuration API.  For instance, `console.json = true` and `console.colorize = true` is no longer a thing.  All those options are configured by way of formatters now.  Writing a translation from 1.x features to 2.x/3.x is not an option as far as I'm concerned.

Furthermore, I think some features got built into some of the custom Transports that have never seen any use.  For instance there was some `preformatting` directives that have never been used to my knowledge.  I would like to dump those legacy behaviors because it only serves to slow things down if it's not useful functionality.

File: lib/loggers/index.js
Comment (mattcollier) [reply lib/loggers/index.js:38]: I don't know who did what when, but the ones we had were definitely in the wrong order.  See: https://github.com/winstonjs/winston#logging-levels

File: lib/loggers/index.js
Comment (mattcollier) [reply lib/loggers/index.js:165]: The `default` formatter here gives us parity with the current console logging we know and love.

File: lib/loggers/index.js
Comment (mattcollier) [reply lib/loggers/index.js:314]: This is some of the legacy stuff I want to flush out.  If it's not being handled by formatters properly now, we can iteratively fix things up.

File: lib/loggers/WorkerTransport.js
Comment (aljones15) [inline lib/loggers/WorkerTransport.js:9]: is there anyway (maybe not now) that we could get this commented with links to the docs for `Windston Transport`. this is so future maintainers can see what API this was creates with vs whatever future version of Winston we implement down the road or other neo-logging solution as Cybernetic screaming unicorns equipped with neural link etc?

**Review by aljones15 — APPROVED** [review]
While this code fails to future proof itself from potential cybernetic or trans-genic logging solutions we will obviously want in the future, this code smells pretty good, looks great, and appears be working. That said I'm not an authority on Winston and did not look at the upgrade guides for Winston when reviewing this PR. I was to busy thinking about cyber-unicorns and such. Logging is a hot and quickly developing space.

**Pull request merged**
